packages/python/plotly/plotly/express/_core.py

```diff
@@ -1341,46 +1420,102 @@ def build_dataframe(args, constructor):
-            # type is similar to the types in other code branches.
-            columns = pd.Index(args["data_frame"].column_names())
-            needs_interchanging = True
-        elif hasattr(args["data_frame"], "to_pandas"):
-            args["data_frame"] = args["data_frame"].to_pandas()
+    if df_provided:
+
+        if nw.dependencies.is_pandas_like_dataframe(args["data_frame"]):
+
+            columns = args["data_frame"].columns  # This can be multi index
+            args["data_frame"] = nw.from_native(args["data_frame"], eager_only=True)
+            is_pd_like = True
+
+        elif nw.dependencies.is_pandas_like_series(args["data_frame"]):
+
+            args["data_frame"] = nw.from_native(
+                args["data_frame"], series_only=True
+            ).to_frame()
             columns = args["data_frame"].columns
-        elif hasattr(args["data_frame"], "toPandas"):
-            args["data_frame"] = args["data_frame"].toPandas()
+            is_pd_like = True
+
+        elif isinstance(
+            data_frame := nw.from_native(
+                args["data_frame"], eager_or_interchange_only=True, strict=False
+            ),
+            nw.DataFrame,
+        ):
+            args["data_frame"] = data_frame
+            needs_interchanging = nw.get_level(data_frame) == "interchange"
             columns = args["data_frame"].columns
-        elif hasattr(args["data_frame"], "to_pandas_df"):
-            args["data_frame"] = args["data_frame"].to_pandas_df()
+
+        elif isinstance(
+            series := nw.from_native(
+                args["data_frame"], series_only=True, strict=False
+            ),
+            nw.Series,
+        ):
+            args["data_frame"] = series.to_frame()
             columns = args["data_frame"].columns
-        else:
-            args["data_frame"] = pd.DataFrame(args["data_frame"])
+
+        elif hasattr(args["data_frame"], "toPandas"):
+            # data_frame is PySpark: it does not support interchange and it is not
+            # integrated in narwhals just yet
+            args["data_frame"] = nw.from_native(
+                args["data_frame"].toPandas(), eager_only=True
+            )
             columns = args["data_frame"].columns
-    elif df_provided:
-        columns = args["data_frame"].columns
+            is_pd_like = True
+
+        else:
+            try:
+                import pandas as pd
+
+                try:
+                    args["data_frame"] = nw.from_native(
+                        pd.DataFrame(args["data_frame"])
+                    )
+                    columns = args["data_frame"].columns
+                    is_pd_like = True
+                except Exception:
+                    msg = f"Unsupported type: {type(args['data_frame'])}"
+                    raise NotImplementedError(msg)
+            except ImportError:
+                msg = (
+                    f"data_frame of type {type(args['data_frame'])} requires Pandas "
```

Comment:
This error message could be a bit misleading since it'll be shown in _all_ the catchall cases, even if we're not sure the `pd.DataFrame` constructer can handle this type of input -- I'd prefer wording along the lines of "Attempting to convert data_frame of type (type) to Pandas, but Pandas is not installed. Convert it to supported dataframe type or install Pandas."